src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final int HALF_DIVIDER_INDEX = 2;
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, findNodeByIndex(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            linkLast(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForAccess(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexForAccess(index);
+        Node<T> x = findNodeByIndex(index);
+        T oldValue = x.item;
+        x.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForAccess(index);
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (indexOf(object) >= 0) {
+            unlink(findNodeByIndex(indexOf(object)));
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public int indexOf(T object) {
+        int index = 0;
+        for (Node<T> x = first; x != null; x = x.next) {
+            if ((size != 0) && (x.item == object || (x.item != null && x.item.equals(object)))) {
+                return index;
+            }
+            index++;
+        }
+        return -1;
+    }
+
+    private T unlink(Node<T> element) {
+        final T item = element.item;
+        final Node<T> prev = element.prev;
+        final Node<T> next = element.next;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            element.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            element.next = null;
+        }
+        element.item = null;
+        size--;
+        return item;
+    }
+
+    private void linkLast(T value) {
+        final Node<T> l = last;
+        final Node<T> newNode = new Node<>(l, value, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
```

Comment:
Be careful with linking Nodes. Make sure that you add mutual links from both sides when you insert a new Node, and that you remove all the links when you remove a Node. Your implementation of `linkLast` correctly handles this.